stubs/xmltodict/xmltodict.pyi

```diff
@@ -0,0 +1,34 @@
+from _typeshed import SupportsRead, SupportsWrite
+from codecs import StreamReaderWriter, StreamWriter
+from collections import OrderedDict
+from typing import IO, Any, AnyStr, Mapping, overload
+
+class ParsingInterrupted(Exception): ...
+
+def parse(
+    xml_input: AnyStr | SupportsRead[bytes],
+    encoding: str | None = ...,
+    expat: Any = ...,
+    process_namespaces: bool = ...,
+    namespace_separator: str = ...,
+    disable_entities: bool = ...,
+    **kwargs: Any,
+) -> OrderedDict[str, Any]: ...
+@overload
+def unparse(
+    input_dict: Mapping[str, Any],
+    output: IO[AnyStr] | StreamWriter | StreamReaderWriter | SupportsWrite[str],
```

Comment:
This is a strange combination of types. The `AnyStr` doesn't make much sense because there's no other `AnyStr` in the signature. The codecs classes are compatible with both `IO[str]` and `SupportsWrite[str]`, so they're redundant. And `IO[str]` is itself redundant with `SupportsWrite[str]`, so you probably don't need that either.

In general we're trying to avoid using `IO` for annotations, in favor of protocols like `SupportsWrite`. I haven't looked at the code for xmltodict, but if it uses this parameter just for `.write()`, then `SupportsWrite[str]` by itself is the right annotation.

Reply:
I copied this from SAX and changed the first two parts of the union type to `IO[AnyStr]` because originally it was either binary or text stream.

Reply:
Here’s the original annotation:
```python
class XMLGenerator(handler.ContentHandler):
    def __init__(
        self,
        out: TextIOBase | RawIOBase | StreamWriter | StreamReaderWriter | SupportsWrite[str] | None = ...,
        encoding: str = ...,
        short_empty_elements: bool = ...,
    ) -> None: ...
```

Reply:
@JelleZijlstra, I’m not aware about the distinction between `IO` and `*IOBase` from the point of view of type annotations; in my code I usually use `IO[…]` with either `bytes` or `str` depending on the context.

Is there any documentation I can check where these conventions are described?

Reply:
https://github.com/python/typing/discussions/829 has some good discussion. I think the SAX annotation is also wrong.

Reply:
I agree with @JelleZijlstra that the SAX annotation looks wrong. (Sorry about that -- it was a very reasonable assumption to think that you could copy this annotation from that annotation!!)

Reply:
So what should I choose? Just `StreamWriter | SupportsWrite[str] | SupportsWrite[bytes]`? It supports writing to text and binary streams, and *also* to objects that have `.write()`.